### PR TITLE
[ORCA] Relax client/server CTYPE encoding requirement

### PR DESCRIPTION
**Pull request by dgkimura**
ORCA represents strings using wide characters. That requires converting
to wide format using `vswprintf()`. However, that API may fail if the
string to covert is incompatible with the set LC_CTYPE character set.
That can happen if the database defined LC_CTYPE character set doesn't
support the client character set.

For example, if the database has LC_CTYPE set to 'C', but the client has
'ko_KR.UTF-8'. In that case, ORCA currently falls back to PLANNER
because converting to wide format fails.

A few possible solutions were considered:
  1) Remove wide characters from ORCA
  2) Use a 'good' LC_CTYPE while calling `vswprintf()`
  3) Return a generic wide character string on failure

Solution 1 is very invasive. Solution 2 is tricky to pick a 'good'
LC_CTYPE. Solution 3 seems the simplest and is the implemented approach.

Solution 3 required re-thinking commit https://github.com/greenplum-db/gpdb/commit/6f0737379c91da56936f6fd7ce166ac15e2abd73. Instead of a fall
back, we allow ORCA use a generic string. Then translate it back to the
actual string during DXLToPlStmt translation. That handles the case
where the offending name is in the columnn project list (see added
regress testcase).

### CODE REVIEW

**Comment by khannaekta** [comment]
Thanks @dgkimura for taking a look into this! 
Returning a generic wide character string on failure maybe a good start but it may not work for some cases like:
```
CREATE DATABASE test WITH LC_COLLATE='C' LC_CTYPE='C' TEMPLATE=template0;
\c test

create table 한글 (col1 int, 컬럼 text) distributed by (컬럼);
insert into 한글 values (1,'abc'), (1,'d'),(1,'ef'),(1,'g'),(1,null);
insert into 한글 values (2,'가나다'),(2,'라'),(2,'마바'),(2,'사'),(2,null);
select * from 한글 order by 1,2;

 col1 | UNKNOWN 
------+---------
    1 | abc
    1 | d
    1 | ef
    1 | g
    1 | 
    2 | 가나다
    2 | 라
    2 | 마바
    2 | 사
    2 | 
(10 rows)
```
Here the column type is returned as an UNKNOWN

**Review by gpopt — COMMENTED** [review]
Can we add more test cases?
- SELECT multiple widecharacters columns
- SELECT queries with wc columns has renamed, or dropped, or added
- UPDATE query
- INSERT into SELECT wc columns...
- SELECT * from (select wc columns ... ) t
- multiple wc columns in CTE

**Comment by dgkimura** [comment]
On Friday, October 27, 2023 10:52 AM gpopt wrote:
> Can we add more test cases?

Yes, thanks a lot for that list!

I added a couple more too for alias and join.  It also showed that the logic needed to be updated a bit.  Incorporated in commit [68da0ae](https://github.com/greenplum-db/gpdb/pull/16619/commits/68da0aefa03fcf1d4ab7c9c4ce4816e2cf485d6a).

File: src/test/regress/sql/gp_locale.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+-- SELECT ALIAS
+EXPLAIN SELECT 안녕세계1 AS 안녕세계1_Alias FROM hi_안녕세계;
+SELECT 안녕세계1 AS 안녕세계1_Alias FROM hi_안녕세계;
+
+-- SUBQUERY
+EXPLAIN SELECT * FROM (SELECT 안녕세계1 FROM hi_안녕세계) t;
+SELECT * FROM (SELECT 안녕세계1 FROM hi_안녕세계) t;
+
+-- CTE
+EXPLAIN WITH cte AS
```
Comment (gpopt) [inline src/test/regress/sql/gp_locale.sql:37]: Can we have another similar test case where the CTE has Japanese column alias? e.g. `WITH cte(わたし, しりつ)`

Comment (dgkimura) [reply src/test/regress/sql/gp_locale.sql]: Yeah, good idea.  Added the Japanese column name.   Also combined the PLANNER/ORCA answer files into one.

File: src/test/regress/sql/gp_locale.sql [hunk 2]
```diff
@@ -0,0 +1,52 @@
+-- ORCA uses functions (e.g. vswprintf) to translation to wide character
+-- format. But those libraries may fail if the current locale cannot handle the
+-- character set. This test checks that even when those libraries fail, ORCA is
+-- still able to generate plans.
+
+--
+-- Create a database that sets the minimum locale
+--
+DROP DATABASE IF EXISTS testdb3;
```
Comment (chrishajas) [inline src/test/regress/sql/gp_locale.sql:9]: Can we name this something like `test_locale`

**Review by chrishajas — APPROVED** [review]
Looks good! And I verified tests work on mac as well

File: src/test/regress/sql/gp_locale.sql [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+-- SELECT ALIAS
+SELECT 안녕세계1 AS 안녕세계1_Alias FROM hi_안녕세계;
+
+-- SUBQUERY
+SELECT * FROM (SELECT 안녕세계1 FROM hi_안녕세계) t;
+
+-- CTE
+WITH cte AS
+(SELECT 안녕세계1, こんにちわ3 FROM hi_안녕세계) SELECT * FROM cte WHERE 안녕세계1 LIKE '안녕세계1%';
```
Comment (gpopt) [inline src/test/regress/sql/gp_locale.sql:47]: Can we have another 2 test cases similar to this?
```sql
WITH cte(안녕세계x, こんにちわx) AS
...
```

and this one:
```
SELECT 안녕세계1 || こんにちわ3 FROM hi_안녕세계;
```

File: src/backend/gporca/libgpos/src/common/clibwrapper.cpp [hunk 4]
```diff
@@ -358,7 +358,7 @@ gpos::clib::Vswprintf(WCHAR *wcstr, SIZE_T max_len, const WCHAR *format,
 	{
 		// Invalid multibyte character encountered. This can happen if the byte sequence does not
 		// match with the server encoding.
-		GPOS_RAISE(CException::ExmaSystem, CException::ExmiIllegalByteSequence);
+		res = swprintf(wcstr, max_len, format, "UNKNOWN");
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpos/src/common/clibwrapper.cpp:365]: Yeah, good call. Added comment.

File: src/test/regress/sql/gp_locale.sql (hunk 2 above)
Comment (dgkimura) [reply src/test/regress/sql/gp_locale.sql]: Done.

File: src/test/regress/sql/gp_locale.sql (hunk 3 above)
Comment (dgkimura) [reply src/test/regress/sql/gp_locale.sql:49]: Yeah, sounds good.  Added.

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 5]
```diff
@@ -5316,6 +5316,51 @@ CTranslatorDXLToPlStmt::ProcessDXLTblDescr(
...
+	{
+		Query *query = (Query *) node;
+
+		return gpdb::WalkExpressionTree(
+			(Node *) query->targetList,
+			(bool (*)()) update_unknown_locale_walker, (void *) context);
+	}
+
+	return gpdb::WalkExpressionTree(
+		node, (bool (*)()) update_unknown_locale_walker, (void *) context);
```
Comment (gpopt) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:5361]: Can you explain why we need to walk through the whole expression tree, given that we only care about the output column names?

Comment (dgkimura) [reply src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:5361]: Thanks for pointing this out!  I forgot to add the testcase.  It's for the case where a target entry is more complex.  For example, without that code the following query would give an UNKNOWN column name:

```sql
select (select こんにちわ3 from hi_안녕세계) from (select 1) as q;
    UNKNOWN    
---------------
 안녕세2 first
(1 row)
```

Whereas with it we get the expected output:

```sql
select (select こんにちわ3 from hi_안녕세계) from (select 1) as q;
  こんにちわ3  
---------------
 안녕세2 first
(1 row)
```

Comment (gpopt) [reply src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:5361]: Beside this test case, can we add another test case where it has 2 level nested subquery in the project column?
select (select (select こんにちわ3 from ....) from hi_안녕세계) from (select 1) as q;

Comment (dgkimura) [reply src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:5361]: Added both tests in last commit

**Comment by dgkimura** [comment]
Merged. Thanks for the reviews!

File: src/backend/gporca/libgpos/src/common/clibwrapper.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (khuddlefish) [reply src/backend/gporca/libgpos/src/common/clibwrapper.cpp:365]: We don't like having the `Hack!` identifier. If it's your implementation then don't call it a "Hack!" It looks bad in the code.

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 6]
```diff
@@ -5420,6 +5465,21 @@ CTranslatorDXLToPlStmt::TranslateDXLProjList(
 				}
 				target_entry->resorigtbl = pteOriginal->resorigtbl;
 				target_entry->resorigcol = pteOriginal->resorigcol;
+
+				// Hack! ORCA represents strings using wide characters. That
```
Comment (khuddlefish) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:5469]: We don't like this "Hack!" word either. Same reason as before

**Pull request merged**